Ffdpacking

avflor opened the pull request:
fixes #934 

The branch contains a new packing algorithm based on binpacking that can support heterogeneous containers and can be used with schedulers like YARN.

objmagic commented:
Thx. Please sign the [CLA](https://engineering.twitter.com/opensource/cla) first.

avflor commented:
I see there are some checkstyle errors which I don't get when building locally and running the tests. Is there any way to run chcekstyle alone and figure out what is wrong?

maosongfu commented:
@avflor 
Try with: bazel build --config=darwin --experimental_action_listener=tools/java:compile_java heron/...

avflor commented:
@maosongfu Thanks! I tried this command and it still works locally. I'm working on Ubuntu but we also tried on a mac and we don't get any checkstyle errors there as well.

objmagic commented:
I checked out your branch and ran

``bazel build --config=darwin --experimental_action_listener=tools/java:compile_java heron/packing/...``

I can see lots of style errors.

maosongfu commented:
@avflor Maybe you need "bazel clean" first? @billonahill may have better ideas on this.

ashvina commented:
Yes. I can see the errors after "bazel clean".

On Mon, Jul 11, 2016 at 1:36 PM, Maosong Fu <notifications@github.com>
wrote:

> @avflor <https://github.com/avflor> Maybe you need "bazel clean" first?
> @billonahill <https://github.com/billonahill> may have better ideas on
> this.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/twitter/heron/pull/1059#issuecomment-231856675>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AJkbjKcI9T_ITc4xiFi585q0ZdwVoBRNks5qUqlagaJpZM4JJrSl>
> .
>

kramasamy commented:
@objmagic - can you cut and paste the checkstyle errors for @avflor?

objmagic commented:
@ashvina are you the same guy as @avflor? and who is @avrilia? what is going on here?

kramasamy commented:
@avflor - one broader comment is to write a documentation about this packing algorithm.

- What does the algorithm do?
- When is it useful and its pros and cons?
- How to enable it?

avflor commented:
@kramasamy I fixed the checkstyle errors. After bazel clean I was able to see them and I checked in the new code. I will write some extra documentation. I have some comments about the algo in the code but I can expand.

@objmagic Thanks for noticing that. There has been some weird configuration issue and apparently GitHub linked my email with an old GitHub account (avrilia). I will fix this and from now on only avflor will be used.

objmagic commented:
@avflor please fix two Java Docs error [here](https://travis-ci.org/twitter/heron/builds/144013684#L1665). You should use ``&lt;`` and ``&gt;``.

kramasamy commented:
@ajorgensen - can you review this PR since you are the closest to write a new packing algorithm?

ajorgensen commented on heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java:
```diff
@@ -0,0 +1,384 @@
+ * the size of ram for an instance is less than the minimal required value.
+ */
+
+public class FFDPacking implements IPacking {
```
`FFD` is a pretty esoteric term. What if we expanded it to `FirstFitDecreasing` or `BinPacking`? I feel like expanding out the name would make it more discoverable.

avflor replied:
Sure, I can do that. Thanks!

ajorgensen commented on heron/packing/src/java/com/twitter/heron/packing/binpacking/FFDPacking.java:
```diff
@@ -0,0 +1,384 @@
+
+    // Take the heron internal container into account and the application master for YARN
+    // scheduler
+    topologyRam += 2 * (instanceRamDefault + DEFAULT_RAM_PADDING_PER_CONTAINER);
```
So this calculation is not generalizable to other schedulers correct? For example Aurora only has 1 extra container, I am not familiar with YARN but it looks like in this case there are 2 extra containers correct?.

ashvina replied:
Good catch. Each topology running on YARN needs one extra container for application master. However the resources allocated to AM container and TMaster container are constant and independent of instanceRam and default padding. Which means the total resources allocated to a topology is a `constant + cost-of-deploying-instances` for both `yarn` and `aurora` clusters.

ajorgensen replied:
I've been thinking about this a little more recently and would be curious what your thoughts are about generic schedulers for all platforms and whether or not we should start splitting some of the at least shell implementations up into separate scheduling technologies (yarn, aurora, mesos, etc). I think that trying to generalize the requirements of all of those platforms into one algorithm might actually be making this more complex than it needs to be and certainly some schedulers will have different trade-offs and be able to make certain optimizations based on how they work. I dont necessarily think we need to do that in this case but I think its worth thinking about how that would work and whether the level effort/complexity that splitting them out into separate scheduler implementations might introduce. @kramasamy I'd be curious of your thoughts as well.

avflor replied:
@ajorgensen We had a discussion with @ashvina about the same topic. The main limitation now is the fact that some schedulers support heterogeneous containers whereas others do not. However, the algorithm should be able to operate on both types of schedulers. The algorithm mainly decides how to pack instances in containers and should provide this packing to the scheduler. The scheduler can then decide to use homogeneous sor heterogeneous containers to represent this packing. I think @ashvina will open an issue on that. In this way, we would be able to use RoundRobin and binpacking on both Aurora and Yarn.

kramasamy commented:
@avflor @ajorgensen @ashvina - I think the consensus to keep the packing algorithms and schedulers orthogonal - the PR for this is here

https://github.com/twitter/heron/pull/1079

kramasamy commented:
Is FFD packing algorithm is good enough to merge?

avflor commented:
I made some updates to address some comments and also changed some configuration parameters. I will check it in tomorrow.

avflor commented:
@ajorgensen Let me know if you have other comments and I will address them tomorrow before I push the current version.

ajorgensen commented:
@avflor I think aside from my initial comments it looks good to me. I have other aspirations to have those default padding values be configurable as well but I don't believe that needs to happen in this pass.

kramasamy commented:
Can we open an issue for tracking?

Sent from my iPhone

> On Jul 15, 2016, at 7:12 AM, Andrew Jorgensen <notifications@github.com> wrote:
> 
> @avflor I think aside from my initial comments it looks good to me. I have other aspirations to have those default padding values be configurable as well but I don't believe that needs to happen in this pass.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ajorgensen commented:
Opened one here: https://github.com/twitter/heron/issues/1084

avflor commented:
@ajorgensen @kramasamy @maosongfu I submitted a revised version of the binpacking algorithm. I noticed when I run in local mode that the number of stream managers is picked based on the topology config. However, the binpacking algorithm decides on its own how many containers to use. The number of stream managers launched should be based on the result of packing algorithm. I enhanced the SchedulerMain.java code to take the value from the PackedPlan. I made the necessary change to the junit tests as well. I also changed the build files so that the Heron client can use the binpacking algorithm. Finally, I made the padding configurable and changed the binpacking algorithm to accommodate it. Please let me know if you have any further comments.

kramasamy commented:
@avflor - excellent. @ajorgensen @maosongfu - if you are fine with the PR, we can merge it.

@avflor - should we check in the documentation as a part of this PR as we discussed earlier or it will go in a separate PR?

avflor commented:
@kramasamy Karthik, what do you think about me submitting the documentation PR as a separate one next week?  I need some more time to look into the tools and the setup. In the meantime, I have comments on how to use the algorithm in the FirstFitDecreasing.java class.

kramasamy commented:
Ok. Please create a github issue and reference it here.

Sent from my iPhone

> On Jul 16, 2016, at 7:44 PM, avflor <notifications@github.com> wrote:
> 
> @kramasamy Karthik, what do you think about me submitting the documentation PR as a separate one next week? I need some more time to look into the tools and the setup. In the meantime, I have comments on how to use the algorithm in the FirstFitDecreasing.java class.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

avflor commented:
First fit decreasing documentation: #1098

ajorgensen commented:
PR looks good to me!

maosongfu commented:
looks good to me. 👍
